### PR TITLE
Update typing-extensions; some 3.12 updates

### PR DESCRIPTION
**Pull request by JelleZijlstra**

### CODE REVIEW

**Comment by JelleZijlstra** [comment]
@rchen152 could you take a look at the failure in https://github.com/python/typeshed/actions/runs/5052177791/jobs/9064782798?pr=10200 ? Feels like it has something to do with `from typing import TypeVar as _TypeVar`.

**Comment by JelleZijlstra** [comment]
Split out the AST changes into #10201, as they're independent from the rest of the changes and less urgent to get in.

File: stdlib/typing.pyi [hunk 1]
```diff
@@ -873,3 +885,23 @@ if sys.version_info >= (3, 10):
     def is_typeddict(tp: object) -> bool: ...
 
 def _type_repr(obj: object) -> str: ...
+
+if sys.version_info >= (3, 12):
+    def override(__arg: _F) -> _F: ...
+
+    class TypeAliasType:
```
Comment (AlexWaygood) [inline stdlib/typing.pyi:892]:
```suggestion
    @_final
    class TypeAliasType:
```

File: stdlib/typing_extensions.pyi [hunk 2]
```diff
@@ -326,5 +340,31 @@ class TypeVarTuple:
...
+
+if sys.version_info >= (3, 12):
+    from collections.abc import Buffer as Buffer
+    from types import get_original_bases as get_original_bases
+    from typing import TypeAliasType as TypeAliasType, override as override
+else:
+    def override(__arg: _F) -> _F: ...
+    def get_original_bases(__cls: type) -> tuple[Any, ...]: ...
+
+    class TypeAliasType:
```
Comment (AlexWaygood) [inline stdlib/typing_extensions.pyi:353]:
```suggestion
    @final
    class TypeAliasType:
```

File: stdlib/_collections_abc.pyi [hunk 3]
```diff
@@ -79,3 +82,7 @@ class dict_items(ItemsView[_KT_co, _VT_co], Generic[_KT_co, _VT_co]):  # undocum
     if sys.version_info >= (3, 10):
         @property
         def mapping(self) -> MappingProxyType[_KT_co, _VT_co]: ...
+
+if sys.version_info >= (3, 12):
+    class Buffer(Protocol):
+        def __buffer__(self, __flags: int) -> memoryview: ...
```
Comment (AlexWaygood) [inline stdlib/_collections_abc.pyi:88]: Huh, I'm surprised this isn't an abstractmethod. It probably should be, right?

**Review by AlexWaygood — APPROVED** [review]

File: stdlib/_collections_abc.pyi (hunk 3 above)
Comment (AlexWaygood) [reply stdlib/_collections_abc.pyi:88]: oh wait, no, it is, I'm confused

Comment (AlexWaygood) [reply stdlib/_collections_abc.pyi:88]:
```suggestion
        @abstractmethod
        def __buffer__(self, __flags: int) -> memoryview: ...
```

File: stdlib/_collections_abc.pyi [hunk 4]
```diff
@@ -79,3 +83,8 @@ class dict_items(ItemsView[_KT_co, _VT_co], Generic[_KT_co, _VT_co]):  # undocum
     if sys.version_info >= (3, 10):
         @property
         def mapping(self) -> MappingProxyType[_KT_co, _VT_co]: ...
+
+if sys.version_info >= (3, 12):
+    class Buffer(Protocol):
```
Comment (AlexWaygood) [inline stdlib/_collections_abc.pyi:88]: Oh wait, this should be `@runtime_checkable` as well, right? (Sorry for the piecemeal review!!)

```suggestion
    @runtime_checkable
    class Buffer(Protocol):
```

File: stdlib/typing.pyi [hunk 5]
```diff
@@ -873,3 +885,23 @@ if sys.version_info >= (3, 10):
...
+            self, name: str, value: Any, *, type_params: tuple[TypeVar | ParamSpec | TypeVarTuple, ...] = ()
+        ) -> None: ...
+        @property
+        def __value__(self) -> Any: ...
+        @property
+        def __type_params__(self) -> tuple[TypeVar | ParamSpec | TypeVarTuple, ...]: ...
+        @property
+        def __parameters__(self) -> tuple[Any, ...]: ...
+        @property
+        def __name__(self) -> str: ...
```
Comment (AlexWaygood) [inline stdlib/typing.pyi:903]: Missing `__module__`

```suggestion
        def __name__(self) -> str: ...
        @property
        def __module__(self) -> str: ...
```

File: stdlib/typing_extensions.pyi [hunk 6]
```diff
@@ -326,5 +340,31 @@ class TypeVarTuple:
...
+            self, name: str, value: Any, *, type_params: tuple[TypeVar | ParamSpec | TypeVarTuple, ...] = ()
+        ) -> None: ...
+        @property
+        def __value__(self) -> Any: ...
+        @property
+        def __type_params__(self) -> tuple[TypeVar | ParamSpec | TypeVarTuple, ...]: ...
+        @property
+        def __parameters__(self) -> tuple[Any, ...]: ...
+        @property
+        def __name__(self) -> str: ...
```
Comment (AlexWaygood) [inline stdlib/typing_extensions.pyi:364]:
```suggestion
        def __name__(self) -> str: ...
        @property
        def __module__(self) -> str: ...
```

File: stdlib/typing.pyi [hunk 7]
```diff
@@ -141,9 +141,21 @@ class TypeVar:
     __constraints__: tuple[Any, ...]
     __covariant__: bool
     __contravariant__: bool
-    def __init__(
-        self, name: str, *constraints: Any, bound: Any | None = None, covariant: bool = False, contravariant: bool = False
-    ) -> None: ...
+    if sys.version_info >= (3, 12):
+        __infer_variance__: bool
```
[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply stdlib/typing.pyi:145]: Yes, nobody should have been writing to those attributes. I suppose we can make them writable on 3.12 if someone complains convincingly enough, but I don't know why you'd want to mutate those attributes.

[the remark this answers is not in the document]
Comment (JelleZijlstra) [reply stdlib/typing.pyi:145]: I will mark them as read-only in typeshed for all versions.

**Review by AlexWaygood — APPROVED** [review]
Looks good!

**Pull request merged**